### ISSUE TITLE
Create LlamaIndex data connector and index

### ISSUE THREAD

**Issue by EtaCassiopeia** [EtaCassiopeia/aicommits-rs#2]
### Context

* LLMs are a phenomenal piece of technology for knowledge generation and reasoning. They are pre-trained on large amounts of publicly available data.
* How do we best augment LLMs with our private data?
* One emerging paradigm is _in-context_ learning (the other is finetuning), where we insert context into the input prompt. That way, we take advantage of the LLM's reasoning capabilities to generate a response.

To perform LLM's data augmentation in a performant, efficient, and cheap manner, we need to solve two components:

* Data Ingestion
* Data Indexing

### Proposed Solution

That's where the LlamaIndex comes in. LlamaIndex is a simple, flexible interface between your external data and LLMs. It provides the following tools in an easy-to-use fashion:

* Offers data connectors to your existing data sources and formats (APIs, PDFs, docs, SQL, etc.)
* Provides indices over your unstructured and structured data for use with LLMs. These indices help to abstract away common boilerplate and pain points for in-context learning:
  * Storing context in an easy-to-access format for prompt insertion.
  * Dealing with prompt limitations (e.g. 4096 tokens for Davinci) when the context is too big.
  * Dealing with text splitting.
* Provides users an interface to query the index (feed in an input prompt) and obtain a knowledge-augmented output.
* Offers you a comprehensive toolset trading off cost and performance

This also can be accomplished by sending a vector of private information with the request:

```Rust
// Create a vector of integers to represent the Git diff
let private_data = [
    "diff --git a/README.md b/README.md",
    "index 5d3e7bb..d0a0984 100644",
    "--- a/README.md",
    "+++ b/README.md",
    "@@ -1,3 +1,4 @@",
    "+# My Project",
    " # My Project",
    " Some description of the project",
    " Another description"
];

// Create a request using the Rust reqwest library
let client = reqwest::Client::new();
let mut response = client
    .post("https://example.com/api/chatgpt/")
    .json(&json!({
        "prompt": "Hello!",
        "private_data": private_data,
    }))
    .send()
    .expect("Failed to send request");

// Read and parse the response
let response_body = response
    .text()
    .expect("Failed to read response body");
let response_data: Value = serde_json::from_str(&response_body)
    .expect("Failed to parse response body");

// Use the response data
println!("Response data: {:?}", response_data);
```

Other alternatives to consider:

- https://github.com/hwchase17/langchain
- https://haystack.deepset.ai/

### Resources:
- https://langchain.readthedocs.io/en/latest/index.html
- https://gpt-index.readthedocs.io/en/latest/
